Document runbook attachment rule and step rule (#91)
firehydrant/terraform-provider-firehydrant: docs/guides/runbooks_attachment_rule.md

```hcl
attachment_rule = jsonencode({
  logic = {
    exists: [
      {
        var: "incident_slack_channel"
      }
    ]
  },
})
```

## Multi condition or

The runbook will attach if a slack incident channel exists or if a microsoft teams channel exists.
```hcl
attachment_rule = jsonencode({
  logic = {
    or: [
      {
        exists: [
          {
            var: "incident_microsoft_teams_channel"
          }
        ]
      },
      {
        exists: [
          {
            var: "incident_slack_channel"
          }
        ]
      }
    ]
  },
})
```

## Multi condition and

The runbook will attach if a slack incident channel exists and a microsoft teams channel exists.
```hcl
attachment_rule = jsonencode({
  logic = {
    and: [
      {
        exists: [
          {
            var: "incident_microsoft_teams_channel"
          }
        ]
      },
      {
        exists: [
          {
            var: "incident_slack_channel"
          }
        ]
      }
    ]
  },
})
```

## Types of conditions

We derive the types of conditions that can be set based on our API
```
https://api.firehydrant.io/v1/fh-attributes/data_bags/system-runbook-attachment-attributes
```

> With the example payload below you can see that we have a list of attributes that can be selected and we can see the `opcode` or operators are able to be used for these attributes. 
> Give these operators we can infer that we can either have a condition where the slack channel does or does not exist.

Reviewer: ```suggestion
In the example payload below you can see that we have a list of attributes that can be selected and we can see the `opcode` or operators that can be used for these attributes. 
Given these operators, we can infer that we can either have a condition where the Slack channel does or does not exist.
```

```json
{
  "name": "system-runbook-attachment-attributes",
  "attributes": [
    {
      "key": "incident_slack_channel",
      "type": "SlackChannel",
      "human_name": "Incident Slack channel"
    }
  ],
  "types": {
    "SlackChannel": {
      "operators": [
        {
          "human_name": "exists",
          "opcode": "exists"
        },
        {
          "human_name": "does not exist",
          "opcode": "does_not_exist"
        }
      ]
    }
  }
}

```

## Value based conditions

> Operators of conditions can be assigned values that are from data saved within different parts of our system. For example the below data we get back from the `system-runbook-attachment-attributes` api shows that we can select multiple operators that take in an `Array` of `IncidentRole` data values. Looking into the `IncidentRole` type we also see values that specify an `async` url that we can use to check and grab the data necessary to fill in the `type`, `value`, and `label` fields required for the user_data.

Reviewer: ```suggestion
Operators of conditions can be assigned values that come from data saved in different parts of our system. For example, the below data we get back from the runbook attachment attributes API shows that we can select multiple operators that take in an array of `IncidentRole` data values. Looking at the `IncidentRole` type, we also see values that specify an `async` url that we can use to get the data necessary to fill in the `type`, `value`, and `label` fields required for the `user_data` attribute.
```

    }
  ],
  "types": {
    "Array[IncidentRole]": {
      "operators": [
        {
          "human_name": "includes any of",
          "opcode": "includes_any",
          "arguments": ["Array[IncidentRole]"]
        },
        {
          "human_name": "includes all of",
          "opcode": "includes_all",
          "arguments": ["Array[IncidentRole]"]
        },
        {
          "human_name": "does not include",
          "opcode": "includes_none_of",
          "arguments": ["Array[IncidentRole]"]
        },
        {
          "human_name": "is empty",
          "opcode": "is_empty"
        }
      ],
      "hints": {
        "related_types": ["IncidentRole"],
        "collection": true
      }
    },
    "IncidentRole": {
      "operators": [
        {
          "human_name": "is",
          "opcode": "eq",
          "arguments": ["IncidentRole"]
        }
      ],
      "values": {
        "async": "/fh-attributes/values/IncidentRole"
      }
    }
  }
}

```

When using the above data attributes we can create something like the rule below where given an `incident_assigned_roles` has the role assigned of `Commander` or `Communication` then attach this runbook.

```hcl

> attachment_rule = jsonencode({
>   logic = {
>     includes_any = [
>       {
>         var = "incident_assigned_roles"
>       },
>       {
>         var = "usr.1"
>       }
>     ]
>   },
>   user_data = {
>     "1" = {
>       type = "Array[IncidentRole]",
>       value = [
>         {
>           type = "IncidentRole",
>           value = "d0c956c1-b285-40a7-89fc-30bb4e5e3b59",
>           label = "Commander"
>         },
>         {
>           type = "IncidentRole",
>           value = "44e8b65d-7682-4279-85f0-ed68cba052cc",
>           label = "Communication"
>         }
>       ],
>     }
>   }
> })

Reviewer: ```suggestion
attachment_rule = jsonencode({
  logic = {
    includes_any = [
      {
        var = "incident_assigned_roles"
      },
      {
        var = "usr.1"
      }
    ]
  }
  user_data = {
    "1" = {
      type = "Array[IncidentRole]"
      value = [
        {
          type = "IncidentRole"
          value = data.firehydrant_incident_role.commander.id
          label = data.firehydrant_incident_role.commander.name
        },
        {
          type = "IncidentRole"
          value = data.firehydrant_incident_role.communications.id
          label = data.firehydrant_incident_role.communications.name
        }
      ]
      label = null
    }
  }
})
```

```

## User data

The user_data object is used to map directly to variables that are held within the logic object. In the below example you can see `var = "user.1"` in the logic object which maps to `user_data["1"]`. As for the key value "1" this can be anything that you would like it to be it just needs to be a unique value within the user_data object that maps to a var in the logic object prepended by `usr.`
```hcl

> attachment_rule = jsonencode({
>   logic = {
>     eq = [
>       {
>         var = "incident_current_milestone",
>       },
>       {
>         var = "usr.1"
>       }
>     ]
>   },
>   user_data = {
>     "1" = {
>       type  = "Milestone",
>       value = "resolved",
>       label = "Resolved"
>     }
>   }
> })

Reviewer: ```suggestion
attachment_rule = jsonencode({
  logic = {
    eq = [
      {
        var = "incident_current_milestone"
      },
      {
        var = "usr.1"
      }
    ]
  }
  user_data = {
    "1" = {
      type  = "Milestone"
      value = "resolved"
      label = "Resolved"
    }
  }
})
```
